css_tweaks: Add generic setting for smaller inpage footnotes

moben opened the pull request:
When fixing the "smaller" versions of in-page footnotes for https://github.com/koreader/koreader/pull/13554 I got annoyed by the duplication. Also, the in-page footnote menu is getting quite big.

I couldn't find any context for why it's set up this way but here is an alternative. `css_tweaks: Generalize smaller inpage footnotes` introduces a single `Smaller in-page footnotes` tweak that affects all enabled footnotes.

I believe a possible downside of this is that it has to be `-cr-hint: late` and needs to be manually sorted by priority relative to other tweaks. But if that were a problem then it would also apply to all the fixup tweaks?

### Screenshot
(also has the changes from https://github.com/koreader/koreader/pull/13554 but that is unrelated, I'll just rebase if one of them is merged)
![Reader_Jonathan Strange and Mr Norrell epub_p1_2025-04-17_015652](https://github.com/user-attachments/assets/1ecd6514-84cb-4f0f-8933-62788f2fba3e)


<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/13613)
<!-- Reviewable:end -->

hius07 commented on frontend/ui/data/onetime_migration.lua:
```diff
@@ -886,5 +886,31 @@ if last_migration_date < 20250405 then
+        global_tweaks["footnote-inpage_wikipedia_smaller"] = nil
+        global_tweaks["footnote-inpage_classic_classnames_smaller"] = nil
+
+        G_reader_settings:saveSetting("style_tweaks", global_tweaks)
```
`global_tweaks` is the table so no need to call save.

Commodore64user commented:
very happy with the removal of all the duplicate settings, although the `extension` stuff could go inside `fix-up` easily. Just add a separator and whack 'em at the bottom.

I do find the names `extend footnotes until next footnote` and `extend footnotes until next header or footnote` a bit too overwhelming. I would suggest something more like `extend footnote selection/content up to next footnote` and `extend footnote selection up to next header`.

unrelated to this PR but, shouldn't that `font size` stuff be a `SpinWidget`?

poire-z commented:
If we can, I'd be all for getting rid of the "smaller" stuff, as it is redundant with choosing a font size (80%) in the "In-page footnote font size" submenu.

But please have a look at (no time myself to re-read all that - and this PR - this evening):
https://github.com/koreader/koreader/pull/5863#issuecomment-1173097714 (where I removed them)
https://github.com/koreader/koreader/pull/5863#issuecomment-1173104241 (where I put them back)
and see if how you manage and migrate that solves all my worries.


> unrelated to this PR but, shouldn't that `font size` stuff be a `SpinWidget`?

These font size items works with each their own id - styletweaks are not dynamic, they are just stored as IDs to static bits of CSS. Having a SpinWidget would need a lot more work to pick the right ID or need some more dynamic CSS spitting, not worth the effort (and I like these menu items, easier that having to play with a SpinWidget).

moben commented:
Pushed an alternative that allows getting rid of "smaller" while keeping all current behavior/combinations in #13639. Not sure if it's worth it though or if this simplification is enough.

poire-z commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -1008,25 +937,27 @@ This tweak can be duplicated as a user style tweak when books contain footnotes
-            conflicts_with = function(id) return util.stringStartsWith(id, "footnote-inpage_") end,
-            global_conflicts_with = function(id) return util.stringStartsWith(id, "footnote-inpage_classic_classnames") end,
-            title = _("In-page classic classname footnotes (smaller)"),
+            id = "smaller_footnote-inpage",
```
We may sort all the settings in settings.reader.lua, so it would be nice if this one started the same as the other.
I think we start with `footnote_inpage_` when targetting elements to be a footnote, and `inpage_footnote_` when adjusting a matched element, so this one could be named `inpage_footnote_smaller`.

poire-z commented:
Looks like you'd be migrating only global tweaks.
What about tweaks set specifically for an already opened book ? The won't be there when that book is reopened ?

moben commented:
Good point, will look into migrating per-book settings as well. (and fix the hint name)

---

In the mean time, thoughts on either of these menu placements? (and of course hint names, I'm bad at those)

<img src="https://github.com/user-attachments/assets/18333266-855c-4776-bf22-354972aedb17" width="45%" />
<img src="https://github.com/user-attachments/assets/90f16d62-c370-4a32-89ca-f42b9ac91c27" width="45%" />

"Smaller footnotes" is just the current 80% setting on top-level footnote blocks, the various sizes would be the same hint with different values. (whereas the "Force" variants are the hint with `inside-inpage-footnote` that applies to all blocks inside a footnote).

Everything in the submenu would conflict to reduce confusion.

Annoyingly, sometimes "small" footnotes are bigger if the body text sets a font smaller than `0.8rem`... I guess the only way to fix that would be a `-cr-scale-font: 0.x` that scales the font after regular `font-size` computation, but I don't want to bloat this PR further.

Commodore64user commented:
I personally hate both those screenshots, but hate the left one a lot less, does it really need to use word “force”?

Frenzie commented:
> I personally hate both those screenshots, but hate the left one a lot less

Forget about how it looks. But what does it *do*? ;-)

I'm not quite sure what this is trying to say. What's the same hint with different values?
> "Smaller footnotes" is just the current 80% setting on top-level footnote blocks, the various sizes would be the same hint with different values. (whereas the "Force" variants are the hint with inside-inpage-footnote that applies to all blocks inside a footnote).

moben commented:
Sorry, I should have elaborated a bit more. I pushed the left version without the wording change.

I just moved the "smaller footnotes" hint, which is 
```
*, autoBoxing {
    -cr-hint: late;
    -cr-only-if: inpage-footnote;
        font-size: 0.8rem !important;
}
```

Into the footnote size submenu. This keeps the same behavior that's in master right now if you select one of the duplicated "smaller" in-page footnote options. (i.e. within footnotes, relative font-sizes work)

I wanted to find some wording that distinguishes this from the footnote font-size tweaks, which override the size of each (block) element in the footnote to e.g. `0.8rem`. (i.e. almost everything in the footnote is forced to the same size)

This has always been confusing in my opinion. If you select the smaller footnotes, what do the footnote font-size hints do? Maybe "override" would be a wording that's more consistent with other tweaks?

In any case, I just split out the duplicated "smaller" part from the footnotes and moved it to the submenu for now (and specified that it's 80%, because all the other size options also specify the size).

@poire-z 

> If we can, I'd be all for getting rid of the "smaller" stuff, as it is redundant with choosing a font size (80%) in the "In-page footnote font size" submenu.
> But please have a look at (no time myself to re-read all that - and this PR - this evening):
> https://github.com/koreader/koreader/pull/5863#issuecomment-1173097714 (where I removed them)
> https://github.com/koreader/koreader/pull/5863#issuecomment-1173104241 (where I put them back)
> and see if how you manage and migrate that solves all my worries.

This doesn't quite get rid of it. I don't think it's possible without fundamentally changing behavior as you were worried about and I think it's valuable to have a tweak that just changes the size at the top level of the footnote (or some way to scale the font-size of a whole block in `crengine`).

This should at least be less cluttered than the duplicated "smaller" footnote hints and correctly does conflicts to make it clear that the font-size tweaks override the "smaller" variant.

---

<img src="https://github.com/user-attachments/assets/9f64f26d-c5b2-4867-8b91-a9f677d2a6e3" width="45%" />
<img src="https://github.com/user-attachments/assets/a5a00694-a1c0-45fe-b1a7-2620faae9ce5" width="45%" />

poire-z commented on frontend/ui/data/onetime_migration.lua:
```diff
@@ -34,15 +34,6 @@ local function drop_fontcache()
     end
 end
 
--- Global settings, https://github.com/koreader/koreader/pull/4945 & https://github.com/koreader/koreader/pull/5655
--- Limit the check to the most recent update. ReaderUI calls this one unconditionally to update docsettings, too.
-if last_migration_date < 20191129 then
-    logger.info("Performing one-time migration for 20191129")
-
-    local SettingsMigration = require("ui/data/settings_migration")
-    SettingsMigration:migrateSettings(G_reader_settings)
```
This one has not been used since 2019 - not sure why. May be it existed before we introduced onetimemigration, which is where migration stuff has been happening :)  @NiLuJe ?
So, dunno if it's a good idea to ressurect/reuse that one, which did/would do other stuff.

NiLuJe replied:
No strong feeling about it, but it probably couldn't hurt to keep the old one as-is, and just add another specific method in `SettingsMigration` for that one ;).

poire-z replied:
onetimemigration also handles global settings. It was introduced by you in April 2021.
settings_migration has had minor stuff added after 2019, but given that it's driven since Apr 2021 by your PR with a last_migration_date < 2019 condition, it has had no effect and use.
I'd be for keeping everything in onetime_migration.

poire-z commented on frontend/ui/data/settings_migration.lua:
```diff
@@ -45,6 +45,57 @@ function SettingsMigration:migrateSettings(config)
     end
 
+    if config:has("style_tweaks") then
+        local tweaks = config:readSetting("style_tweaks")
+
+        if tweaks then
```
(Uneeded blank line.)

poire-z commented on frontend/ui/data/settings_migration.lua:
```diff
@@ -45,6 +45,57 @@ function SettingsMigration:migrateSettings(config)
+        local tweaks = config:readSetting("style_tweaks")
+
+        if tweaks then
+            --            base hint
+            --
+            --          | n | t | f |
+            --       ---+---+---+---+
+            -- small  n | n | t | f |
+            -- hint   t | t | t | t |
+            --        f | f | t | f |
```
What do `n`, `t` and `f` stand for ? I really can't say :/ so it needs a comment.

Some minutes later: Oh, nil, true and false ! you may as well write them fully :)

Some minutes later: what does `hint` means ? What's the last unamed line, the resut ?

moben replied:
I hope it's clearer now. The two input dimensions are the tweak setting for e.g. _classname footnotes_ and _smaller classname footnotes_, the fields are the result after the migration.

I found it helpful to write down all the cases especially because the meaning differs a bit between global and per-book settings.

poire-z replied:
Oh, "base hint" written on 2 lines :)
"hint" feels like not the best word, I would just use "note".

moben replied:
I changed it to hopefully be unambiguous and also make it clear that this is per footnote type. (diff above is outdated though)

For some reason my brain keeps replacing _"tweak"_ with _"hint"_, but I think I fixed that in all comments (probably because of `-cr-hint`)...

poire-z commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -1045,6 +956,29 @@ This tweak can be duplicated as a user style tweak when books contain footnotes
             local sub_table = {
                 title = _("In-page footnote font size"),
+                {
+                    id = "inpage_footnote_font-size_smaller",
+                    conflicts_with = function(id) return util.stringStartsWith(id, "inpage_footnote_font-size_") end,
+                    title = _("Smaller footnotes (80%)"),
+                    description = _([[
+Decrease size of in-page footnotes.]]),
+                    css = [[
+body[name="notes"] > section {
+    -cr-only-if: fb2-document;
+        font-size: 0.75rem;
+}
+body[name="comments"] > section {
+    -cr-only-if: fb2-document;
+        font-size: 0.85rem;
+}
+*, autoBoxing {
+    -cr-hint: late;
+    -cr-only-if: -fb2-document inpage-footnote;
+        font-size: 0.8rem !important;
+}
+                    ]],
+                    separator = true,
+                },
```
OK, what I disliked in the other PR, I'm ok with having it only in this "somehow legacy" tweak.
May be it could have a more verbose description, mentionning it's limited (no inheritance, so possibly no effect - the reasong why you qualitied the others with "force" all along).

moben replied:
I'm honestly not 100% sure if I like this better or if it's better to keep the fb2 footnote size tweak separate. My main concern was cleaning up the top-level footnote menu and I got a bit carried away.

As you noted, it prevents some combinations like having larger epub and smaller fb2.
Any preference on your side for reducing menu items vs keeping fb2 size separate? (as in master, or also moving it inside font size)

poire-z replied:
Dunno, not sure I get what the options are :).
I don't mint _not_ reducing menu stuff for FB2 and having them separated (FB2 is useful to user from a few countries only, 98% of users won't ever hear about FB2).
I'd rather not have the font size % duplicated in the FB2 submenu.

moben replied:
I reverted the changes to fb2, incorporating that into the migration would have been tricky anyway.

It now just removes the "smaller" top-level variants and moves them to the footnote font-size submenu.

poire-z commented:
Ok, I prefer this one.

> I couldn't find any context for why it's set up this way

Initially, we only had the pairs "normal" and "smaller", which were not "forcing": no inheritance to sub elements (because otherwise it would have killed any superscript in the footnote), so possibly having no effect if there's nested block elements and the publisher sets a font size on the sub element.

Later, I added support for `-cr-only-if: -inline`, which allowed me to target all nested block elements (with `inside-inpage-footnote`), while preserving inline (superscripts) relative font size.

The 2nd behaviour is probably the best, and could be the only one.
I would have been ok killing the 1st one, and just forcing the 2nd to users, they would probably not notice - and those who do can make their own book style tweak when it has unwanted effect.

But as you are keeping this 1st behaviour (that I'm willing to call legacy) buried inside the "Footnote font size" submenu, I'm quite fine with this PR.

No idea if the migration will handle all cases, or prevent some combinations (smaller FB2 but normal EPUB ? probably rarely used), so trusting you.

moben commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -879,26 +872,12 @@ body[name="comments"] > title {
             id = "footnote-inpage_epub",
             conflicts_with = function(id) return util.stringStartsWith(id, "footnote-inpage_") end,
-            global_conflicts_with = function(id) return util.stringStartsWith(id, "footnote-inpage_epub") end,
+            global_conflicts_with = false,
```
by the way, is it intentional that this lets you set multiple of these footnote tweaks as global defaults, even if they conflict? You can do that on master and I've preserved the behavior, but I always found it strange. As soon as you then adjust them in book-specific settings you can only keep one, so it's very inconsistent.

Frenzie replied:
Indeed, the question is what you mean by conflict exactly. Setting all the footnote classes as default is very reasonable for example.

moben commented:
> No idea if the migration will handle all cases, or prevent some combinations (smaller FB2 but normal EPUB ? probably rarely used), so trusting you.

I tested many different combinations, but this obviously limits possibilities compared to what's in `master`. e.g. you can't have regular size EPUB footnotes and smaller classname footnotes as default. (or in a single book if such a book exists)

As you say, probably rarely used though and the newer footnote font-size tweaks already work the same way and affect all footnotes, so this is probably better anyway.

poire-z commented:
Fine with the latest comments and changes.

Remaining, from my previous review comments:

https://github.com/koreader/koreader/pull/13613#issuecomment-2819268550
> Looks like you'd be migrating only global tweaks.
What about tweaks set specifically for an already opened book ? The won't be there when that book is reopened ?

https://github.com/koreader/koreader/pull/13613#discussion_r2060820188

> onetimemigration also handles global settings. It was introduced by you in April 2021.
settings_migration has had minor stuff added after 2019, but given that it's driven since Apr 2021 by your PR with a last_migration_date < 2019 condition, it has had no effect and use.
I'd be for keeping everything in onetime_migration.

https://github.com/koreader/koreader/pull/13613#discussion_r2060785397

> May be it could have a more verbose description, mentionning it's limited (no inheritance, so possibly no effect - the reasong why you qualitied the others with "force" all along).

moben commented:
`settings_migration` handles global and per-book settings. I switched to using that and this branch now handles both correctly (that's where the `false` cases in the migration table come from, global tweaks are only `true` or `nil` as far as I understand)

> since Apr 2021 by your PR with a last_migration_date < 2019 condition, [settings_migration] has had no effect and use.

I don't think that's quite correct. `settings_migration` is actually called in [readerui.lua](https://github.com/koreader/koreader/blob/master/frontend/apps/reader/readerui.lua#L129-L130) each time any book is opened and actually re-runs the same migration each time. I didn't want to touch the whole migration here because it would be a bigger change, but I'm happy to merge them into one place that handles both global and per-book (and actually also respects/sets `last_migration_date` for per-book settings)

poire-z commented:
Oh, you mean it's alright, and the correct place, to have per-book (docsettings) settings migration ?
And that if we hadn't touched that file for years, it's because we just haven't had the need to migrate past opened book settings ? Strange that we haven't had the need, but ok.
Ok then.

poire-z commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -1045,6 +978,21 @@ This tweak can be duplicated as a user style tweak when books contain footnotes
+                    conflicts_with = function(id) return util.stringStartsWith(id, "inpage_footnote_font-size_") end,
+                    title = _("Smaller footnotes (80%)"),
+                    description = _([[
+Decrease size of in-page footnotes. This may have limited/no effect, e.g. if it's overwritten by publisher styles.]]),
```
It will find the footnote container and apply it.
But if the publisher has set a font-size on a sub-block of that container, this font-size will not be overridden by ours (set on the parent, not inherited).
So, may have no apparent effect.

The other (added at a later date) tweaks use `* { -cr-only-if: inside-inpage-footnote }` and will match ALL children and will override their font-size.
That's why I mentionned previously this one could be considered legacy and this could be mentionned - but @moben seems to want to be soft about it :)

moben replied:
I'm not _trying_ to be soft about it, I'm just bad at writing these descriptions :smile:

moben replied:
I made the wording stronger, maybe it's better now?

poire-z replied:
Fine with me :)

moben commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -24,8 +24,9 @@ local T = require("ffi/util").template
-DEFAULT_GLOBAL_STYLE_TWEAKS["footnote-inpage_epub_smaller"] = true
+DEFAULT_GLOBAL_STYLE_TWEAKS["footnote-inpage_epub"] = true
 DEFAULT_GLOBAL_STYLE_TWEAKS["footnote-inpage_fb2"] = true
+DEFAULT_GLOBAL_STYLE_TWEAKS["inpage_footnote_font-size_smaller"] = true
```
Another thing that could be done is switching this default to `inpage_footnote_font-size_80`.

That would apply to new installs, whereas existing users get the "legacy" tweak `inpage_footnote_font-size_smaller` via the migration.

Maybe a good idea if it's strongly recommended to use the newer tweaks as @poire-z suggests. Then we'd also be more likely to get reports if these newer tweaks are worse in some situation. After some time one could migrate everyone from the legacy tweak (if wanted).

poire-z replied:
As you feel.

Frenzie commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -1045,6 +978,21 @@ This tweak can be duplicated as a user style tweak when books contain footnotes
+                    conflicts_with = function(id) return util.stringStartsWith(id, "inpage_footnote_font-size_") end,
+                    title = _("Smaller footnotes (80%, legacy)"),
+                    description = _([[
+Decrease size of in-page footnotes. This may have limited/no apparent effect, e.g. if it's overwritten by publisher styles. The "footnote font size" style tweaks are recommended instead.]]),
```
```suggestion
Decrease size of in-page footnotes. This may have no effect if it's overwritten by publisher styles. The "footnote font size" style tweaks are recommended instead.]]),
```

poire-z commented:
"This branch has conflicts that must be resolved". Letting you handle this.

poire-z commented on frontend/ui/data/onetime_migration.lua:
```diff
@@ -12,7 +12,7 @@ local util = require("util")
 
 -- Date at which the last migration snippet was added
-local CURRENT_MIGRATION_DATE = 20250405
+local CURRENT_MIGRATION_DATE = 20250417
```
Can you update this date to the day you'll do it :)
